scanpointgenerator/core/compoundgenerator.py

```diff
@@ -21,236 +23,315 @@ def __init__(self, generators, excluders, mutators):
+                generators[generators.index(gen_1)] = new_gen1
+                generators[generators.index(gen_2)] = new_gen2
+                excluders.remove(rect)
+
+        for generator in generators:
+            generator.produce_points()
+            self.axes_points.update(generator.points)
+            self.axes_points_lower.update(generator.points_lower)
+            self.axes_points_upper.update(generator.points_upper)
+            self.num *= generator.num
 
-            # If innermost generator, use bounds
-            reverse, sub_point = self._get_sub_point(
-                num_point_sets - 1, point_num)
-            point.positions.update(sub_point.positions)
-            if reverse:  # Swap bounds if reversing
-                point.upper.update(sub_point.lower)
-                point.lower.update(sub_point.upper)
+            dim = {"size":generator.num,
+                "axes":list(generator.axes),
+                "generators":[generator],
+                "masks":[],
+                "tile":1,
+                "repeat":1,
+                "alternate":generator.alternate_direction}
+            self.dimensions.append(dim)
+
+        for excluder in excluders:
+            axis_1, axis_2 = excluder.scannables
+            # ensure axis_1 is "outer" axis (if separate generators)
+            gen_1 = [g for g in generators if axis_1 in g.axes][0]
+            gen_2 = [g for g in generators if axis_2 in g.axes][0]
+            gen_diff = generators.index(gen_1) \
+                - generators.index(gen_2)
+            if gen_diff < -1 or gen_diff > 1:
+                raise ValueError(
+                    "Excluders must be defined on axes that are adjacent in " \
+                        "generator order")
+            if gen_diff == 1:
+                gen_1, gen_2 = gen_2, gen_1
+                axis_1, axis_2 = axis_2, axis_1
+                gen_diff = -1
+
+
+            #####
+            # first check if region spans two dimensions - merge if so
+            #####
+            dim_1 = [i for i in self.dimensions if axis_1 in i["axes"]][0]
+            dim_2 = [i for i in self.dimensions if axis_2 in i["axes"]][0]
+            dim_diff = self.dimensions.index(dim_1) \
+                - self.dimensions.index(dim_2)
+            if dim_diff < -1 or dim_diff > 1:
+                raise ValueError(
+                    "Excluders must be defined on axes that are adjacent in " \
+                        "generator order")
+            if dim_diff == 1:
+                dim_1, dim_2 = dim_2, dim_1
+                dim_diff = -1
+            if dim_1["alternate"] != dim_2["alternate"] \
+                    and dim_1 is not self.dimensions[0]:
+                raise ValueError(
+                    "Generators tied by regions must have the same " \
+                            "alternate_direction setting")
+            # merge "inner" into "outer"
+            if dim_diff == -1:
```

Comment:
... Or is dim_diff == 0 the case for spiral and lissajous where one generator already contains the axes for a region? Does the list of dims have to match the list of excluders?